src/kessel/grpc/auth.py

```diff
@@ -0,0 +1,92 @@
+import datetime
+
+import google.auth.credentials
+import google.auth.transport.requests
+import requests
+from oauthlib.oauth2 import BackendApplicationClient
+from requests_oauthlib import OAuth2Session
+
+
+class ClientCredentials(google.auth.credentials.Credentials):
+    """
+    Client credentials class for handling the OAuth 2.0 Client Credentials flow.
+
+    Integrates with the google-auth and requests-oauthlib library to fetch an access token
+    from a specified issuer url with automatic refreshing.
+    """
+
+    def __init__(
+        self,
+        issuer_url: str,
+        client_id: str,
+        client_secret: str,
+    ):
+        """
+        Initializes the ClientCredentials.
+
+        Args:
+            issuer_url: The issuer URL of the OAuth 2.0 provider.
+            client_id: The client ID for the application.
+            client_secret: The client secret for the application.
+        """
+        super().__init__()
+        self._issuer_url = issuer_url
+        self._client_id = client_id
+        self._client_secret = client_secret
+
+        self._token_url = self._discover_token_endpoint(issuer_url)
```

Comment:
I wonder if we should do this lazily rather than on init?

I lean towards lazily, with an explicit (and optional) "initialize()" (or similarly named) method which an application can call when it is ready (e.g. a django app's [ready callback](https://docs.djangoproject.com/en/5.2/ref/applications/#django.apps.AppConfig.ready)), but not required, because we'll also just call it on the first request.

I say that because I generally find it surprising that merely instantiating an object requires heavy construction, like network calls.

Reply:
👍 I like the idea of lazily instantiating things, I had done the oidc discovery part last and had originally had it input the direct `token_url` so the setup here was more of a drop in.